{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "pd.set_option('display.max_colwidth', -1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "abstracts = pd.read_csv('/home/grace/Downloads/ChemProt_Corpus/chemprot_training/chemprot_training/chemprot_training_abstracts.tsv', sep='\\t', header=None)\n",
    "abstracts.columns = ['pmid', 'title', 'abstract']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pmid</th>\n",
       "      <th>title</th>\n",
       "      <th>abstract</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>16357751</td>\n",
       "      <td>Selective costimulation modulators: a novel approach for the treatment of rheumatoid arthritis.</td>\n",
       "      <td>T cells have a central role in the orchestration of the immune pathways that contribute to the inflammation and joint destruction characteristic of rheumatoid arthritis (RA). The requirement for a dual signal for T-cell activation and the construction of a fusion protein that prevents engagement of the costimulatory molecules required for this activation has led to a new approach to RA therapy. This approach is mechanistically distinct from other currently used therapies; it targets events early rather than late in the immune cascade, and it results in immunomodulation rather than complete immunosuppression. The fusion protein abatacept is a selective costimulation modulator that avidly binds to the CD80/CD86 ligands on an antigen-presenting cell, resulting in the inability of these ligands to engage the CD28 receptor on the T cell. Abatacept dose-dependently reduces T-cell proliferation, serum concentrations of acute-phase reactants, and other markers of inflammation, including the production of rheumatoid factor by B cells. Recent studies have provided consistent evidence that treatment with abatacept results in a rapid onset of efficacy that is maintained over the course of treatment in patients with inadequate response to methotrexate and anti-tumor necrosis factor therapies. This efficacy includes patient-centered outcomes and radiographic measurement of disease progression. Abatacept has also demonstrated a very favorable safety profile to date. This article reviews the rationale for this therapeutic approach and highlights some of the recent studies that demonstrate the benefits obtained by using abatacept. This clinical experience indicates that abatacept is a significant addition to the therapeutic armamentarium for the management of patients with RA.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>14967461</td>\n",
       "      <td>Emerging role of epidermal growth factor receptor inhibition in therapy for advanced malignancy: focus on NSCLC.</td>\n",
       "      <td>Combination chemotherapy regimens have emerged as the standard approach in advanced non-small-cell lung cancer. Meta-analyses have demonstrated a 2-month increase in median survival after platinum-based therapy vs. best supportive care, and an absolute 10% improvement in the 1-year survival rate. Just as importantly, cytotoxic therapy has produced benefits in symptom control and quality of life. Newer agents, including the taxanes, vinorelbine, gemcitabine, and irinotecan, have expanded our therapeutic options in the treatment of advanced non-small-cell lung cancer. Despite their contributions, we have reached a therapeutic plateau, with response rates seldom exceeding 30-40% in cooperative group studies and 1-year survival rates stable between 30% and 40%. It is doubtful that substituting one agent for another in various combinations will lead to any further improvement in these rates. The thrust of current research has focused on targeted therapy, and epidermal growth factor receptor inhibition is one of the most promising clinical strategies. Epidermal growth factor receptor inhibitors currently under investigation include the small molecules gefitinib (Iressa, ZD1839) and erlotinib (Tarceva, OSI-774), as well as monoclonal antibodies such as cetuximab (IMC-225, Erbitux). Agents that have only begun to undergo clinical evaluation include CI-1033, an irreversible pan-erbB tyrosine kinase inhibitor, and PKI166 and GW572016, both examples of dual kinase inhibitors (inhibiting epidermal growth factor receptor and Her2). Preclinical models have demonstrated synergy for all these agents in combination with either chemotherapy or radiotherapy, leading to great enthusiasm regarding their ultimate contribution to lung cancer therapy. However, serious clinical challenges persist. These include the identification of the optimal dose(s); the proper integration of these agents into popular, established cytotoxic regimens; and the selection of the optimal setting(s) in which to test these compounds. Both gefitinib and erlotinib have shown clinical activity in pretreated, advanced non-small-cell lung cancer, but placebo-controlled randomized Phase III studies evaluating gefitinib in combination with standard cytotoxic therapy, to our chagrin, have failed to demonstrate a survival advantage compared with chemotherapy alone.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>23468099</td>\n",
       "      <td>Effects of chronic social defeat stress on behavior and choline acetyltransferase, 78-kDa glucose-regulated protein, and CCAAT/enhancer-binding protein (C/EBP) homologous protein in adult mice.</td>\n",
       "      <td>RATIONALE: Social defeat stress induces physiological and behavioral symptoms, including anxiety, anhedonia, immune deficits, and altered expression of key brain genes. OBJECTIVES: The present study investigated the effects of social defeat stress on the behaviors and expressions of Chat, Grp78, and chop in the brains of adult mice. METHODS: Adult mice were divided into susceptible and unsusceptible groups after 10Â days of social defeat stress. In experiment 1, behavioral tests were conducted, and brains were processed for Western blotting at day 27 after stress. In experiment 2, social avoidance tests were conducted, and brains were processed for Western blotting at day 12 after stress. RESULTS: The results indicate decreased and increased locomotion and anxiety behavior in all defeated mice. Decrease in social interaction, increased immobility, and impaired memory performance were only observed in susceptible mice. A decrease in the Chat level at days 12 and 27 was noted in the prefrontal cortex (PFC), amygdala (Amyg), and dorsal hippocampus (HIP) in defeated mice. The expression levels of Grp78 and chop measured on days 12 and 27 were significantly greater in the Amyg of susceptible mice. In the PFC and HIP, defeated mice displayed different patterns in the levels of Grp78 and chop expressions measured on days 12 and 27. CONCLUSIONS: The present study demonstrated that chronic social defeat stress in mice produces stress-related behaviors. Different response patterns were noted for Grp78 and chop expression among the groups in terms of brain regions and time-course effects.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>23293962</td>\n",
       "      <td>Hepatocyte growth factor activator inhibitor type 2 (HAI-2) modulates hepcidin expression by inhibiting the cell surface protease matriptase-2.</td>\n",
       "      <td>Matriptase-2, a recently identified cell surface protease, is the key enzyme of iron homoeostasis modulating the expression of the liver peptide hormone hepcidin. HAI (hepatocyte growth factor activator inhibitor) types 1 and 2 (HAI-1 and HAI-2 respectively) have been shown to inhibit the close homologue, i.e. matriptase. By co-expressing matriptase-2 and the inhibitor HAI-2 we have identified HAI-2 displaying high inhibitory potential against matriptase-2 at the cell surface as well as in conditioned medium. Accordingly, complex formation between matriptase-2 and HAI-2 was demonstrated by isolation of the complex via immobilizing either HAI-2 or matriptase-2 from lysates and conditioned medium of co-expressing cells. Furthermore, HAI-2 indirectly influences the expression of the hepcidin-encoding gene HAMP. The inhibitor abrogates the matriptase-2-mediated suppression of HAMP expression, presumably by inhibiting the supposed potential of matriptase-2 to cleave membrane-bound HJV (haemojuvelin). Taken together, the results of the present study have characterized HAI-2 as an inhibitor of matriptase-2 that modulates the synthesis of hepcidin and provides new insights into the regulatory mechanism of iron homoeostasis, with clinical importance for a treatment of iron overload diseases.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7678677</td>\n",
       "      <td>Alprenolol and bromoacetylalprenololmenthane are competitive slowly reversible antagonists at the beta 1-adrenoceptors of rat left atria.</td>\n",
       "      <td>We studied the effects of alprenolol and bromoacetylalprenololmenthane (BAAM) on rat left atria. Alprenolol and BAAM at 10(-7), 3 x 10(-7), and 10(-6) M inhibited the cardiac stimulation response slightly, which is indicative of membrane-stabilizing activity independent of beta-adrenoceptor blockade. This membrane-stabilizing activity was readily reversible. Alprenolol and BAAM also caused surmountable antagonism of isoprenaline responses, and this beta 1-adrenoceptor antagonism was slowly reversible. Inhibition of the isoprenaline responses with alprenolol and BAAM at 10(-6) M was at equilibrium after 60 min, which is indicative of reversible antagonism. We conclude that alprenolol and BAAM are competitive slowly reversible beta 1-adrenoceptor antagonists on rat left atria.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       pmid  \\\n",
       "0  16357751   \n",
       "1  14967461   \n",
       "2  23468099   \n",
       "3  23293962   \n",
       "4  7678677    \n",
       "\n",
       "                                                                                                                                                                                               title  \\\n",
       "0  Selective costimulation modulators: a novel approach for the treatment of rheumatoid arthritis.                                                                                                     \n",
       "1  Emerging role of epidermal growth factor receptor inhibition in therapy for advanced malignancy: focus on NSCLC.                                                                                    \n",
       "2  Effects of chronic social defeat stress on behavior and choline acetyltransferase, 78-kDa glucose-regulated protein, and CCAAT/enhancer-binding protein (C/EBP) homologous protein in adult mice.   \n",
       "3  Hepatocyte growth factor activator inhibitor type 2 (HAI-2) modulates hepcidin expression by inhibiting the cell surface protease matriptase-2.                                                     \n",
       "4  Alprenolol and bromoacetylalprenololmenthane are competitive slowly reversible antagonists at the beta 1-adrenoceptors of rat left atria.                                                           \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           abstract  \n",
       "0  T cells have a central role in the orchestration of the immune pathways that contribute to the inflammation and joint destruction characteristic of rheumatoid arthritis (RA). The requirement for a dual signal for T-cell activation and the construction of a fusion protein that prevents engagement of the costimulatory molecules required for this activation has led to a new approach to RA therapy. This approach is mechanistically distinct from other currently used therapies; it targets events early rather than late in the immune cascade, and it results in immunomodulation rather than complete immunosuppression. The fusion protein abatacept is a selective costimulation modulator that avidly binds to the CD80/CD86 ligands on an antigen-presenting cell, resulting in the inability of these ligands to engage the CD28 receptor on the T cell. Abatacept dose-dependently reduces T-cell proliferation, serum concentrations of acute-phase reactants, and other markers of inflammation, including the production of rheumatoid factor by B cells. Recent studies have provided consistent evidence that treatment with abatacept results in a rapid onset of efficacy that is maintained over the course of treatment in patients with inadequate response to methotrexate and anti-tumor necrosis factor therapies. This efficacy includes patient-centered outcomes and radiographic measurement of disease progression. Abatacept has also demonstrated a very favorable safety profile to date. This article reviews the rationale for this therapeutic approach and highlights some of the recent studies that demonstrate the benefits obtained by using abatacept. This clinical experience indicates that abatacept is a significant addition to the therapeutic armamentarium for the management of patients with RA.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    \n",
       "1  Combination chemotherapy regimens have emerged as the standard approach in advanced non-small-cell lung cancer. Meta-analyses have demonstrated a 2-month increase in median survival after platinum-based therapy vs. best supportive care, and an absolute 10% improvement in the 1-year survival rate. Just as importantly, cytotoxic therapy has produced benefits in symptom control and quality of life. Newer agents, including the taxanes, vinorelbine, gemcitabine, and irinotecan, have expanded our therapeutic options in the treatment of advanced non-small-cell lung cancer. Despite their contributions, we have reached a therapeutic plateau, with response rates seldom exceeding 30-40% in cooperative group studies and 1-year survival rates stable between 30% and 40%. It is doubtful that substituting one agent for another in various combinations will lead to any further improvement in these rates. The thrust of current research has focused on targeted therapy, and epidermal growth factor receptor inhibition is one of the most promising clinical strategies. Epidermal growth factor receptor inhibitors currently under investigation include the small molecules gefitinib (Iressa, ZD1839) and erlotinib (Tarceva, OSI-774), as well as monoclonal antibodies such as cetuximab (IMC-225, Erbitux). Agents that have only begun to undergo clinical evaluation include CI-1033, an irreversible pan-erbB tyrosine kinase inhibitor, and PKI166 and GW572016, both examples of dual kinase inhibitors (inhibiting epidermal growth factor receptor and Her2). Preclinical models have demonstrated synergy for all these agents in combination with either chemotherapy or radiotherapy, leading to great enthusiasm regarding their ultimate contribution to lung cancer therapy. However, serious clinical challenges persist. These include the identification of the optimal dose(s); the proper integration of these agents into popular, established cytotoxic regimens; and the selection of the optimal setting(s) in which to test these compounds. Both gefitinib and erlotinib have shown clinical activity in pretreated, advanced non-small-cell lung cancer, but placebo-controlled randomized Phase III studies evaluating gefitinib in combination with standard cytotoxic therapy, to our chagrin, have failed to demonstrate a survival advantage compared with chemotherapy alone.  \n",
       "2  RATIONALE: Social defeat stress induces physiological and behavioral symptoms, including anxiety, anhedonia, immune deficits, and altered expression of key brain genes. OBJECTIVES: The present study investigated the effects of social defeat stress on the behaviors and expressions of Chat, Grp78, and chop in the brains of adult mice. METHODS: Adult mice were divided into susceptible and unsusceptible groups after 10Â days of social defeat stress. In experiment 1, behavioral tests were conducted, and brains were processed for Western blotting at day 27 after stress. In experiment 2, social avoidance tests were conducted, and brains were processed for Western blotting at day 12 after stress. RESULTS: The results indicate decreased and increased locomotion and anxiety behavior in all defeated mice. Decrease in social interaction, increased immobility, and impaired memory performance were only observed in susceptible mice. A decrease in the Chat level at days 12 and 27 was noted in the prefrontal cortex (PFC), amygdala (Amyg), and dorsal hippocampus (HIP) in defeated mice. The expression levels of Grp78 and chop measured on days 12 and 27 were significantly greater in the Amyg of susceptible mice. In the PFC and HIP, defeated mice displayed different patterns in the levels of Grp78 and chop expressions measured on days 12 and 27. CONCLUSIONS: The present study demonstrated that chronic social defeat stress in mice produces stress-related behaviors. Different response patterns were noted for Grp78 and chop expression among the groups in terms of brain regions and time-course effects.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               \n",
       "3  Matriptase-2, a recently identified cell surface protease, is the key enzyme of iron homoeostasis modulating the expression of the liver peptide hormone hepcidin. HAI (hepatocyte growth factor activator inhibitor) types 1 and 2 (HAI-1 and HAI-2 respectively) have been shown to inhibit the close homologue, i.e. matriptase. By co-expressing matriptase-2 and the inhibitor HAI-2 we have identified HAI-2 displaying high inhibitory potential against matriptase-2 at the cell surface as well as in conditioned medium. Accordingly, complex formation between matriptase-2 and HAI-2 was demonstrated by isolation of the complex via immobilizing either HAI-2 or matriptase-2 from lysates and conditioned medium of co-expressing cells. Furthermore, HAI-2 indirectly influences the expression of the hepcidin-encoding gene HAMP. The inhibitor abrogates the matriptase-2-mediated suppression of HAMP expression, presumably by inhibiting the supposed potential of matriptase-2 to cleave membrane-bound HJV (haemojuvelin). Taken together, the results of the present study have characterized HAI-2 as an inhibitor of matriptase-2 that modulates the synthesis of hepcidin and provides new insights into the regulatory mechanism of iron homoeostasis, with clinical importance for a treatment of iron overload diseases.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           \n",
       "4  We studied the effects of alprenolol and bromoacetylalprenololmenthane (BAAM) on rat left atria. Alprenolol and BAAM at 10(-7), 3 x 10(-7), and 10(-6) M inhibited the cardiac stimulation response slightly, which is indicative of membrane-stabilizing activity independent of beta-adrenoceptor blockade. This membrane-stabilizing activity was readily reversible. Alprenolol and BAAM also caused surmountable antagonism of isoprenaline responses, and this beta 1-adrenoceptor antagonism was slowly reversible. Inhibition of the isoprenaline responses with alprenolol and BAAM at 10(-6) M was at equilibrium after 60 min, which is indicative of reversible antagonism. We conclude that alprenolol and BAAM are competitive slowly reversible beta 1-adrenoceptor antagonists on rat left atria.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "abstracts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1020, 3)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "abstracts.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "entities = pd.read_csv('/home/grace/Downloads/ChemProt_Corpus/chemprot_training/chemprot_training/chemprot_training_entities.tsv', sep='\\t', header=None)\n",
    "# entity term (CHEMICAL, GENE-Y, GENE-N=mention that cannot be normalized to a database identifier)\n",
    "entities.columns = ['pmid', 'entity_num', 'entity_term', 'start', 'end', 'mention']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pmid</th>\n",
       "      <th>entity_num</th>\n",
       "      <th>entity_term</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>mention</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11319232</td>\n",
       "      <td>T1</td>\n",
       "      <td>CHEMICAL</td>\n",
       "      <td>242</td>\n",
       "      <td>251</td>\n",
       "      <td>acyl-CoAs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>11319232</td>\n",
       "      <td>T2</td>\n",
       "      <td>CHEMICAL</td>\n",
       "      <td>1193</td>\n",
       "      <td>1201</td>\n",
       "      <td>triacsin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11319232</td>\n",
       "      <td>T3</td>\n",
       "      <td>CHEMICAL</td>\n",
       "      <td>1441</td>\n",
       "      <td>1448</td>\n",
       "      <td>sucrose</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11319232</td>\n",
       "      <td>T4</td>\n",
       "      <td>CHEMICAL</td>\n",
       "      <td>1637</td>\n",
       "      <td>1652</td>\n",
       "      <td>triacylglycerol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11319232</td>\n",
       "      <td>T5</td>\n",
       "      <td>CHEMICAL</td>\n",
       "      <td>1702</td>\n",
       "      <td>1711</td>\n",
       "      <td>acyl-CoAs</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       pmid entity_num entity_term  start   end          mention\n",
       "0  11319232  T1         CHEMICAL    242    251   acyl-CoAs      \n",
       "1  11319232  T2         CHEMICAL    1193   1201  triacsin       \n",
       "2  11319232  T3         CHEMICAL    1441   1448  sucrose        \n",
       "3  11319232  T4         CHEMICAL    1637   1652  triacylglycerol\n",
       "4  11319232  T5         CHEMICAL    1702   1711  acyl-CoAs      "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "entities.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_x = abstracts.loc[abstracts.pmid==11319232]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "x_t = sample_x.title.tolist()[0]\n",
    "x = sample_x.abstract.tolist()[0]\n",
    "# x = x.replace('(', ' ( ').replace(')', ' ) ')\n",
    "x = x_t + ' ' + x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1836"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "267"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_arr = x.split(' ')\n",
    "len(x_arr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>content</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acyl-CoA</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>synthetase</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>isoforms</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1,</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4,</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      content  length\n",
       "0  Acyl-CoA    8     \n",
       "1  synthetase  10    \n",
       "2  isoforms    8     \n",
       "3  1,          2     \n",
       "4  4,          2     "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_df = pd.DataFrame(x_arr)\n",
    "x_df.columns = ['content']\n",
    "x_df['length'] = x_df.content.str.len()\n",
    "x_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/grace/.local/lib/python3.5/site-packages/pandas/core/indexing.py:189: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  self._setitem_with_indexer(indexer, value)\n"
     ]
    }
   ],
   "source": [
    "x_df['start_index'] = 0\n",
    "x_df['t_length'] = x_df.length\n",
    "for ind, row in x_df.iterrows():\n",
    "    if ind >0:\n",
    "        x_df['t_length'].iloc[ind] = row['length']+x_df.iloc[ind-1]['t_length']+1\n",
    "        x_df['start_index'].iloc[ind] = x_df['t_length'].iloc[ind] - row['length']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>content</th>\n",
       "      <th>length</th>\n",
       "      <th>t_length</th>\n",
       "      <th>start_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>257</th>\n",
       "      <td>metabolic</td>\n",
       "      <td>9</td>\n",
       "      <td>1763</td>\n",
       "      <td>1754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>pathways</td>\n",
       "      <td>8</td>\n",
       "      <td>1772</td>\n",
       "      <td>1764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>through</td>\n",
       "      <td>7</td>\n",
       "      <td>1780</td>\n",
       "      <td>1773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>different</td>\n",
       "      <td>9</td>\n",
       "      <td>1790</td>\n",
       "      <td>1781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>ACS</td>\n",
       "      <td>3</td>\n",
       "      <td>1794</td>\n",
       "      <td>1791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>isoforms</td>\n",
       "      <td>8</td>\n",
       "      <td>1803</td>\n",
       "      <td>1795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263</th>\n",
       "      <td>in</td>\n",
       "      <td>2</td>\n",
       "      <td>1806</td>\n",
       "      <td>1804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>264</th>\n",
       "      <td>unique</td>\n",
       "      <td>6</td>\n",
       "      <td>1813</td>\n",
       "      <td>1807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265</th>\n",
       "      <td>subcellular</td>\n",
       "      <td>11</td>\n",
       "      <td>1825</td>\n",
       "      <td>1814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266</th>\n",
       "      <td>locations.</td>\n",
       "      <td>10</td>\n",
       "      <td>1836</td>\n",
       "      <td>1826</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         content  length  t_length  start_index\n",
       "257  metabolic    9       1763      1754       \n",
       "258  pathways     8       1772      1764       \n",
       "259  through      7       1780      1773       \n",
       "260  different    9       1790      1781       \n",
       "261  ACS          3       1794      1791       \n",
       "262  isoforms     8       1803      1795       \n",
       "263  in           2       1806      1804       \n",
       "264  unique       6       1813      1807       \n",
       "265  subcellular  11      1825      1814       \n",
       "266  locations.   10      1836      1826       "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_df.tail(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/grace/.local/lib/python3.5/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pmid</th>\n",
       "      <th>entity_num</th>\n",
       "      <th>entity_term</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>mention</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>11319232</td>\n",
       "      <td>T11</td>\n",
       "      <td>GENE-N</td>\n",
       "      <td>1212</td>\n",
       "      <td>1215</td>\n",
       "      <td>ACS</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>11319232</td>\n",
       "      <td>T21</td>\n",
       "      <td>GENE-N</td>\n",
       "      <td>1791</td>\n",
       "      <td>1794</td>\n",
       "      <td>ACS</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>11319232</td>\n",
       "      <td>T30</td>\n",
       "      <td>GENE-N</td>\n",
       "      <td>197</td>\n",
       "      <td>200</td>\n",
       "      <td>ACS</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>11319232</td>\n",
       "      <td>T32</td>\n",
       "      <td>GENE-N</td>\n",
       "      <td>840</td>\n",
       "      <td>843</td>\n",
       "      <td>ACS</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>11319232</td>\n",
       "      <td>T34</td>\n",
       "      <td>GENE-N</td>\n",
       "      <td>947</td>\n",
       "      <td>950</td>\n",
       "      <td>ACS</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>11319232</td>\n",
       "      <td>T37</td>\n",
       "      <td>GENE-N</td>\n",
       "      <td>1081</td>\n",
       "      <td>1084</td>\n",
       "      <td>ACS</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        pmid entity_num entity_term  start   end mention  length\n",
       "10  11319232  T11        GENE-N      1212   1215  ACS     3     \n",
       "20  11319232  T21        GENE-N      1791   1794  ACS     3     \n",
       "29  11319232  T30        GENE-N      197    200   ACS     3     \n",
       "31  11319232  T32        GENE-N      840    843   ACS     3     \n",
       "33  11319232  T34        GENE-N      947    950   ACS     3     \n",
       "36  11319232  T37        GENE-N      1081   1084  ACS     3     "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y = entities.loc[entities.pmid==11319232]\n",
    "y['length'] = y.mention.str.len()\n",
    "y.tail()\n",
    "y[y.mention=='ACS']\n",
    "# for ind, row in y.sort_values(by=['length'], ascending=False).iterrows():\n",
    "#     print(row['mention'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# x_df[(x_df.start_index==1081) & (x_df.t_length==1084)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Acyl-CoA synthetase isoforms 1, 4, and 5\n",
      "0 40\n",
      "      content  length  t_length  start_index label\n",
      "0  Acyl-CoA    8       8         0            O   \n",
      "1  synthetase  10      19        9            O   \n",
      "2  isoforms    8       28        20           O   \n",
      "3  1,          2       31        29           O   \n",
      "4  4,          2       34        32           O   \n",
      "5  and         3       38        35           O   \n",
      "6  5           1       40        39           O   \n",
      "acyl-CoA synthetase\n",
      "176 195\n",
      "       content  length  t_length  start_index label\n",
      "26  acyl-CoA    8       184       176          O   \n",
      "27  synthetase  10      195       185          O   \n",
      "N-ethylmaleimide\n",
      "790 806\n",
      "Empty DataFrame\n",
      "Columns: [content, length, t_length, start_index, label]\n",
      "Index: []\n",
      "triacylglycerol\n",
      "1637 1652\n",
      "             content  length  t_length  start_index label\n",
      "242  triacylglycerol  15      1652      1637         O   \n",
      "Troglitazone\n",
      "898 910\n",
      "Empty DataFrame\n",
      "Columns: [content, length, t_length, start_index, label]\n",
      "Index: []\n",
      "Triacsin C\n",
      "1012 1022\n",
      "      content  length  t_length  start_index label\n",
      "145  Triacsin  8       1020      1012         O   \n",
      "acyl-CoAs\n",
      "242 251\n",
      "      content  length  t_length  start_index label\n",
      "37  acyl-CoAs  9       251       242          O   \n",
      "acyl-CoAs\n",
      "1702 1711\n",
      "       content  length  t_length  start_index label\n",
      "250  acyl-CoAs  9       1711      1702         O   \n",
      "acyl-CoA\n",
      "176 184\n",
      "     content  length  t_length  start_index     label\n",
      "26  acyl-CoA  8       184       176          B-GENE-N\n",
      "Acyl-CoA\n",
      "0 8\n",
      "    content  length  t_length  start_index     label\n",
      "0  Acyl-CoA  8       8         0            B-GENE-N\n",
      "triacsin\n",
      "1193 1201\n",
      "Empty DataFrame\n",
      "Columns: [content, length, t_length, start_index, label]\n",
      "Index: []\n",
      "sucrose\n",
      "1441 1448\n",
      "     content  length  t_length  start_index label\n",
      "210  sucrose  7       1448      1441         O   \n",
      "ACS4\n",
      "1615 1619\n",
      "    content  length  t_length  start_index label\n",
      "237  ACS4    4       1619      1615         O   \n",
      "ACS1\n",
      "1244 1248\n",
      "Empty DataFrame\n",
      "Columns: [content, length, t_length, start_index, label]\n",
      "Index: []\n",
      "ACS4\n",
      "1250 1254\n",
      "Empty DataFrame\n",
      "Columns: [content, length, t_length, start_index, label]\n",
      "Index: []"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/grace/.local/lib/python3.5/site-packages/pandas/core/indexing.py:189: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  self._setitem_with_indexer(indexer, value)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "ACS5\n",
      "1260 1264\n",
      "    content  length  t_length  start_index label\n",
      "179  ACS5    4       1264      1260         O   \n",
      "ACS4\n",
      "1369 1373\n",
      "    content  length  t_length  start_index label\n",
      "197  ACS4    4       1373      1369         O   \n",
      "ACS5\n",
      "1402 1406\n",
      "Empty DataFrame\n",
      "Columns: [content, length, t_length, start_index, label]\n",
      "Index: []\n",
      "ACS1\n",
      "1496 1500\n",
      "    content  length  t_length  start_index label\n",
      "221  ACS1    4       1500      1496         O   \n",
      "ACSs\n",
      "372 376\n",
      "   content  length  t_length  start_index label\n",
      "56  ACSs    4       376       372          O   \n",
      "ACS4\n",
      "1505 1509\n",
      "    content  length  t_length  start_index label\n",
      "223  ACS4    4       1509      1505         O   \n",
      "ACS1\n",
      "1606 1610\n",
      "    content  length  t_length  start_index label\n",
      "235  ACS1    4       1610      1606         O   \n",
      "ACS5\n",
      "525 529\n",
      "Empty DataFrame\n",
      "Columns: [content, length, t_length, start_index, label]\n",
      "Index: []\n",
      "ACS1\n",
      "509 513\n",
      "Empty DataFrame\n",
      "Columns: [content, length, t_length, start_index, label]\n",
      "Index: []\n",
      "ACS4\n",
      "515 519\n",
      "Empty DataFrame\n",
      "Columns: [content, length, t_length, start_index, label]\n",
      "Index: []\n",
      "ACS1\n",
      "531 535\n",
      "   content  length  t_length  start_index label\n",
      "77  ACS1    4       535       531          O   \n",
      "ACS4\n",
      "655 659\n",
      "   content  length  t_length  start_index label\n",
      "92  ACS4    4       659       655          O   \n",
      "ACS5\n",
      "705 709\n",
      "    content  length  t_length  start_index label\n",
      "101  ACS5    4       709       705          O   \n",
      "ACS4\n",
      "824 828\n",
      "Empty DataFrame\n",
      "Columns: [content, length, t_length, start_index, label]\n",
      "Index: []\n",
      "ACS4\n",
      "921 925\n",
      "    content  length  t_length  start_index label\n",
      "131  ACS4    4       925       921          O   \n",
      "ACS1\n",
      "1056 1060\n",
      "    content  length  t_length  start_index label\n",
      "152  ACS1    4       1060      1056         O   \n",
      "ACS4\n",
      "1065 1069\n",
      "Empty DataFrame\n",
      "Columns: [content, length, t_length, start_index, label]\n",
      "Index: []\n",
      "ACS\n",
      "1791 1794\n",
      "    content  length  t_length  start_index label\n",
      "261  ACS     3       1794      1791         O   \n",
      "ACS\n",
      "197 200\n",
      "Empty DataFrame\n",
      "Columns: [content, length, t_length, start_index, label]\n",
      "Index: []\n",
      "ACS\n",
      "840 843\n",
      "    content  length  t_length  start_index label\n",
      "118  ACS     3       843       840          O   \n",
      "ACS\n",
      "947 950\n",
      "    content  length  t_length  start_index label\n",
      "134  ACS     3       950       947          O   \n",
      "ACS\n",
      "1212 1215\n",
      "    content  length  t_length  start_index label\n",
      "171  ACS     3       1215      1212         O   \n",
      "ACS\n",
      "1081 1084\n",
      "    content  length  t_length  start_index label\n",
      "156  ACS     3       1084      1081         O   \n"
     ]
    }
   ],
   "source": [
    "x_df['label'] = 'O'\n",
    "y = y.sort_values(by=['length'], ascending=False)\n",
    "for ind, row in y.iterrows():\n",
    "    print(row['mention'])\n",
    "    print(row['start'], row['end'])\n",
    "    mentions = x_df[(x_df.start_index>=row['start']) & (x_df.t_length<=row['end'])]\n",
    "    print(mentions)\n",
    "    ii = 0\n",
    "    for ind_m, m in mentions.iterrows():\n",
    "        if x_df['label'].iloc[ind_m]=='O':\n",
    "            if ii==0: \n",
    "                x_df['label'].iloc[ind_m] = 'B-'+row['entity_term']\n",
    "            else:\n",
    "                x_df['label'].iloc[ind_m] = 'I-'+row['entity_term']\n",
    "        ii += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "x_df.to_csv('./test.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
